artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/impl/PagingStoreImpl.java

```diff
@@ -228,11 +228,58 @@ private void configureSizeMetric() {
       size.setMax(maxSize, maxSize, maxMessages, maxMessages);
    }
 
+   private boolean validateNewSettings(final AddressSettings addressSettings) {
+
+      Long newPageLimitBytes = addressSettings.getPageLimitBytes();
+
+      if (newPageLimitBytes != null && newPageLimitBytes < 0) {
+         newPageLimitBytes = null;
+      }
+
+      int newPageSize = storageManager.getAllowedPageSize(addressSettings.getPageSizeBytes());
+
+      if (newPageLimitBytes != null && newPageSize > 0) {
+
+         if (newPageSize > newPageLimitBytes) {
+            ActiveMQServerLogger.LOGGER.pageSettingsFailedApply(storeName, addressSettings, "pageLimitBytes is smaller than pageSize. It should allow at least one page file");
+            return false;
+         }
+
+         long newEstimatedMaxPages = newPageLimitBytes / newPageSize;
+
+         long existingNumberOfPages;
+         if (!running) {
+            try {
+               checkFileFactory();
+               List<String> files = this.fileFactory.listFiles("page");
+               existingNumberOfPages = files.size();
+            } catch (Exception e) {
+               ActiveMQServerLogger.LOGGER.pageSettingsFailedApply(storeName, addressSettings, "failed to get existing page files with exception: " + e);
+               return false;
+            }
+         } else {
+            existingNumberOfPages = this.numberOfPages;
+         }
+
+         if (existingNumberOfPages > newEstimatedMaxPages) {
+            ActiveMQServerLogger.LOGGER.pageSettingsFailedApply(storeName, addressSettings, "estimated max pages " + newEstimatedMaxPages + " (pageLimitBytes / pageSize) is less than current number of pages " + existingNumberOfPages);
+            return false;
+         }
```

Comment:
Per my other comments, I now question this entirely (not the log message, but the return false). If someone wants to set a limit, that is lower than the current usage, its not clear to me that is a reason not to set it. They can consume those thigns to get back below it, and untilt hen it would act as if at the limit and stop more being added, which is as close to the desired behaviour as its possible to get. Failing to set anything might just mean usage keeps increasing beyond what it already is, i.e the exact opposite of what they want to happen.

Reply:
I think your idea is more friendly to the users. But that'll change the current implementation of it. According to current impl, the practical usage of pageLimitBytes would be a hard rule, i.e. in that case the user has to consume the messages until the current pages is below the expected new limit, that apply the new limit. Based on your idea the new limit can be set at anytime during broker running. something worth to do. Anyway we need to document that behavior.

Reply:
The diff makes it seem like you are adding that 'you cant set a limit lower than what is in use' behaviour here in this commit? You are saying that the existing code already does that, where is that happening? Also feels awkward to have it happening in 2 places without apparent relation.

I was actually more interested in the appearance of a 'shut broker down, decrease limit, start up...fail new validation, get no settings applied at all' issue that I commented on around the test, but runtime behaviour is also a key issue.

Reply:
I've updated the PR and a bit document. Please review see if I missed anything. Thanks!